'zabbix_action' module minor improvements

vzabawski opened the pull request:
##### SUMMARY
- Zabbix 4.2 API has 4 new types of operators, zabbix actions module needs to be updated in order to support them.
- Update mandatory fields in accordance with Zabbix documentation.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
zabbix_action

##### ADDITIONAL INFORMATION
Current implementation doesn't allow to create a potentially correct rule because of failed validation on Ansible's side. For example, I can't create an action with empty "operations" field. Zabbix API allows to do that because there are cases when "operations" field is empty but "operations_recovery" field contains some data.

vzabawski committed:
Adds minor improvements to 'zabbix_actions' module

Changes:

    1. Update mandatory arguments in accordance with Zabbix API.
       Only esc_period, eventsource and name fields are mandatory.

       Reference: https://www.zabbix.com/documentation/3.0/manual/api/reference/action/object

    2. Add default values to argument options in order to prevent "TypeError: 'NoneType' object is not iterable" error.

    3. Add new condition operators (introduced in Zabbix 4.0 and 4.2)

    References:
      - https://www.zabbix.com/documentation/4.2/manual/api/reference/action/object#action_filter_condition
      - https://www.zabbix.com/documentation/4.0/manual/api/reference/action/object#action_filter_condition

    4. Update module documentation

vzabawski commented:
Seems like all Shippable steps are failed with `Couldn't find remote ref pull`. Can anyone tip me how to fix this?

D3DeFi reviewed (commented):
Thanks for identifying and sending fix for this @vzabawski ! 

Can you add some information to description of `operator` option regarding these 4 options being supported only with zbx >= 4.2?

D3DeFi commented on lib/ansible/modules/monitoring/zabbix/zabbix_action.py:
```diff
@@ -127,6 +127,7 @@
                 description:
                     - Condition operator.
                     - When I(type) is set to C(time_period), the choices are C(in), C(not in).
+                    - 'matches', 'does not match', 'Yes' and 'No' condition operators work only with >= Zabbix 4.0
```
Failing because of:
```suggestion
                    - C(matches), C(does not match), C(Yes) and C(No) condition operators work only with >= Zabbix 4.0
```

D3DeFi reviewed (approved):
LGTM

vzabawski commented:
Thanks for the assistance!

rubentsirunyan reviewed (approved):
Thank you @vzabawski!
shipit

vzabawski commented:
I'm sorry for making you start over your approval process, guys, but I've discovered an inconsistency between Zabbix documentation and 'zabbix_action' module implementation that blocks my work, so I've added a commit that fixes it. :sweat_smile:

D3DeFi commented on lib/ansible/modules/monitoring/zabbix/zabbix_action.py:
```diff
@@ -64,6 +64,7 @@
         description:
             - Default operation step duration. Must be greater than 60 seconds. Accepts seconds, time unit with suffix and user macro.
         default: '60'
+        required: true
```
There are 2 places where `esc_period` is defined in argument_spec, but none of it has `required=True`, so this change would make and inconsistency between module documentation and actual argument definition.

D3DeFi replied:
This is a valid change as it is specified in [Zabbix API action object documentation](https://www.zabbix.com/documentation/3.2/manual/api/reference/action/object) as well. Please adjust [this line](https://github.com/ansible/ansible/blob/62b3fec7055cfb1eef6b95cace8dd324a16fd97b/lib/ansible/modules/monitoring/zabbix/zabbix_action.py#L1677) to reflect this documentation change - `required=True`.

D3DeFi commented on lib/ansible/modules/monitoring/zabbix/zabbix_action.py:
```diff
@@ -794,15 +800,15 @@ def _construct_parameters(self, **kwargs):
+            'r_shortdata': kwargs.get('recovery_default_subject'),
+            'ack_longdata': kwargs.get('acknowledge_default_message'),
+            'ack_shortdata': kwargs.get('acknowledge_default_subject'),
+            'operations': kwargs.get('operations'),
```
Every module argument defaults to `None` if not overridden via `default=something` in `argument_spec` so this change may be unnecessary at all, but I am be missing something here that you see.

vzabawski replied:
Maybe you're right and this change is not unnecessary. The reason why I made it was because if I submit and empty field that is defined as `kwargs['something']`, an exception will be risen.
Seems like Zabbix API need any type of operations to be present in request: operations/recovery operations/acknowledge operations. I'm not that good at Python, so I've decided not to check if any of those fields are present in config and just let Zabbix API to handle errors.
Could you advise me how this problem can be solved? The only thing I wanted to fix is validation that makes 'operations' field mandatory although mandatory fields are conditional and only one of them must present.
So, I'm open for suggestions from more experienced developers. :)

D3DeFi replied:
@vzabawski can you paste the exception you are getting please? Your change may be valid, but I am still having trouble understanding it fully.

@rubentsirunyan what's your opinion on this?

D3DeFi replied:
This may get overriden with what I've mentioned in other [comment](https://github.com/ansible/ansible/pull/58379/files#r301533237) as this module wouldn't let you use itself without `esc_period` argument.

For the other arguments (changed on lines 804 - 811) change from `kwargs['XYZ']` to `kwargs.get('XYZ')` will still return `None` as default.

vzabawski replied:
Code:
```
- name: Test action
  zabbix_action:
    server_url: "{{ zabbix_url }}"
    login_user: "{{ zabbix_login }}"
    login_password: "{{ zabbix_password }}"
    name: "Test action"
    event_source: trigger
    esc_period: 3600
    state: present
    status: enabled

    eval_type: "and"

    conditions:
      - type: trigger_name
        operator: "like"
        value: "memory"
    
    acknowledge_operations:
      - type: send_message
        media_type: 'Email'
        send_to_users:
          - 'Admin'
```
Error:
```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: TypeError: 'NoneType' object is not iterable
fatal: [127.0.0.1]: FAILED! => {"changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"/home/valery/.ansible/tmp/ansible-tmp-1562674167.5-10943360351739/AnsiballZ_zabbix_action.py\", line 114, in <module>\n    _ansiballz_main()\n  File \"/home/valery/.ansible/tmp/ansible-tmp-1562674167.5-10943360351739/AnsiballZ_zabbix_action.py\", line 106, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/home/valery/.ansible/tmp/ansible-tmp-1562674167.5-10943360351739/AnsiballZ_zabbix_action.py\", line 49, in invoke_module\n    imp.load_module('__main__', mod, module, MOD_DESC)\n  File \"/tmp/ansible_zabbix_action_payload_gTRlch/__main__.py\", line 2070, in <module>\n  File \"/tmp/ansible_zabbix_action_payload_gTRlch/__main__.py\", line 2061, in main\n  File \"/tmp/ansible_zabbix_action_payload_gTRlch/__main__.py\", line 1104, in construct_the_data\nTypeError: 'NoneType' object is not iterable\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}
```

Info:
```
valery@valery-VirtualBox:~$ ansible --version
ansible 2.8.1
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/valery/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.15+ (default, Nov 27 2018, 23:36:35) [GCC 7.3.0]
```

vzabawski replied:
Now I install patched version:
```
valery@valery-VirtualBox:~$ ansible --version
ansible 2.9.0.dev0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/valery/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.15+ (default, Nov 27 2018, 23:36:35) [GCC 7.3.0]
```
Output (partial):
```
TASK [Test action] *************************************************************
changed: [127.0.0.1]

PLAY RECAP *********************************************************************
127.0.0.1                  : ok=4    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```
The action appears in Zabbix.
Verbose output:
https://gist.github.com/vzabawski/f9f42c484a60e05bfdf84642cfc06c34

rubentsirunyan replied:
```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: TypeError: 'NoneType' object is not iterable
fatal: [127.0.0.1]: FAILED! => {"changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"/home/valery/.ansible/tmp/ansible-tmp-1562674167.5-10943360351739/AnsiballZ_zabbix_action.py\", line 114, in <module>\n    _ansiballz_main()\n  File \"/home/valery/.ansible/tmp/ansible-tmp-1562674167.5-10943360351739/AnsiballZ_zabbix_action.py\", line 106, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/home/valery/.ansible/tmp/ansible-tmp-1562674167.5-10943360351739/AnsiballZ_zabbix_action.py\", line 49, in invoke_module\n    imp.load_module('__main__', mod, module, MOD_DESC)\n  File \"/tmp/ansible_zabbix_action_payload_gTRlch/__main__.py\", line 2070, in <module>\n  File \"/tmp/ansible_zabbix_action_payload_gTRlch/__main__.py\", line 2061, in main\n  File \"/tmp/ansible_zabbix_action_payload_gTRlch/__main__.py\", line 1104, in construct_the_data\nTypeError: 'NoneType' object is not iterable\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}
```

Actually, this error has nothing to do with this change (`kwargs[]` or `kwargs.get()`). So @D3DeFi is right, there is no need to use `kwargs.get()`, as not specified arguments default to `None`.
The reason why your patch does not throw this error is this change you made:
https://github.com/ansible/ansible/blob/62b3fec7055cfb1eef6b95cace8dd324a16fd97b/lib/ansible/modules/monitoring/zabbix/zabbix_action.py#L1116-L1117
Without it, if you try pass `None` to the method it tries to iterate over it which obviously fails with `TypeError: 'NoneType' object is not iterable`

I think defaulting all the arguments that have `type=list` to `[]` is a prettier solution to this issue.
What do you think?

D3DeFi replied:
Agree, explicit things are usually more readable and cleaner solution. Btw, can you drop changelog fragment to [this folder as well](https://github.com/ansible/ansible/tree/devel/changelogs/fragments) please?

No need to document that we are changing defaults from `None` to `[]`, but information like: "`esc_period` is now required to reflect actual Zabbix API call" and "support for new condition operators (x, y, z) added in Zabbix 4.2" is a good info to have in release changelog, when this hits stable release.

Thank you both for all the work so far

vzabawski replied:
@D3DeFi So, could you please clarify what should be done from my side? I guess I need to revert changes made to arguments and leave the following fragment:
```
if operations is None: 
return None 
```

Am I right? Btw, should I squash all commits into one?

D3DeFi replied:
1. `esc_period` in `argument_spec` should have `required=True` instead of `False` to match your documentation change
2. changes on lines 803 - 811 from your patch should be reverted (`kwargs[]` vs `kwargs.get()`)
3. argument options `conditions` (line 1690), `operations` (line 1704), `recovery_operations` (line 1812) and `acknowledge_operations` (line 1896) should have `default=[]` parameter as they are `type=list`
4. changelog fragment should be added to ansible-repo/changelog/fragments directory (see my earlier comment in this thread)

D3DeFi replied:
And yes, it would be cool if you could squash all of the commits into one!

vzabawski commented:
@D3DeFi I've updated my PR in accordance with your comments. I had to remove default value from `esc_period`, otherwise CI was failing the test since an argument can't be mandatory and have a default value at the same time. Also I've removed default value from the documentation.

D3DeFi commented:
Awesome, thanks! @rubentsirunyan as an author of this module, are you satisfied with proposed changes, can we vote for merge?

rubentsirunyan commented on lib/ansible/modules/monitoring/zabbix/zabbix_action.py:
```diff
@@ -1107,6 +1112,8 @@ def construct_the_data(self, operations):
         """
         constructed_data = []
+        if operations is None:
+            return None
```
As we already have `default=[]` for operations, there is no need for this block anymore.

rubentsirunyan reviewed (changes requested):
Please add `default=[]` for `recovery_operations` and `acknowledge_operations` too.

vzabawski commented:
@rubentsirunyan Seems like `recovery_operations` and `acknowledge_operations` already have `default=[]`.
https://github.com/ansible/ansible/blob/fe8eceaeb498a20f6714a2d14eb2acfe3a002e55/lib/ansible/modules/monitoring/zabbix/zabbix_action.py#L1816
https://github.com/ansible/ansible/blob/fe8eceaeb498a20f6714a2d14eb2acfe3a002e55/lib/ansible/modules/monitoring/zabbix/zabbix_action.py#L1900

vzabawski commented:
@rubentsirunyan I've removed code blocks you've mentioned. Is everything okay now?

rubentsirunyan reviewed (approved):
Looks good to me.
Thank you very much for your time @vzabawski!

D3DeFi commented:
Thank you both for your work
LGTM

rubentsirunyan commented:
shipit
